src/win/tcp.c

```diff
@@ -49,29 +49,99 @@ static int uv__tcp_nodelay(uv_tcp_t* handle, SOCKET socket, int enable) {
 }
 
 
-static int uv__tcp_keepalive(uv_tcp_t* handle, SOCKET socket, int enable, unsigned int delay) {
+/*
+ * Check if Windows version is 10.0.16299 (Windows 10, version 1709) or later.
+ */
+static int minimal_windows10_version1709(void) {
+  OSVERSIONINFOW os_info;
+  if (!pRtlGetVersion)
+    return 0;
+  pRtlGetVersion(&os_info);
+  if (os_info.dwMajorVersion < 10)
+    return 0;
+  if (os_info.dwMajorVersion > 10)
+    return 1;
+  if (os_info.dwMinorVersion > 0)
+    return 1;
+  return os_info.dwBuildNumber >= 16299;
+}
+
+
+static int uv__tcp_keepalive(uv_tcp_t* handle,
+                             SOCKET socket,
+                             int on,
+                             unsigned int idle,
+                             unsigned int intvl,
+                             unsigned int cnt) {
   if (setsockopt(socket,
                  SOL_SOCKET,
                  SO_KEEPALIVE,
-                 (const char*)&enable,
-                 sizeof enable) == -1) {
+                 (const char*)&on,
+                 sizeof on) == -1) {
     return WSAGetLastError();
   }
 
-  if (!enable)
+  if (!on)
     return 0;
 
-  if (delay < 1)
+  if (idle < 1 || intvl < 1 || cnt < 1)
     return UV_EINVAL;
 
-  if (setsockopt(socket,
-                 IPPROTO_TCP,
-                 TCP_KEEPALIVE,
-                 (const char*)&delay,
-                 sizeof delay) == -1) {
-    return WSAGetLastError();
+  /* Windows 10, version 1709 (build 10.0.16299) and later require second units
+   * for TCP keepalive options. */
+  if (minimal_windows10_version1709()) {
+    if (setsockopt(socket,
+                  IPPROTO_TCP,
+                  TCP_KEEPIDLE,
+                  (const char*)&idle,
+                  sizeof idle) == -1) {
+      return WSAGetLastError();
+    }
+
+    if (setsockopt(socket,
+                  IPPROTO_TCP,
+                  TCP_KEEPINTVL,
+                  (const char*)&intvl,
+                  sizeof intvl) == -1) {
+      return WSAGetLastError();
+    }
+
+    if (setsockopt(socket,
+                  IPPROTO_TCP,
+                  TCP_KEEPCNT,
+                  (const char*)&cnt,
+                  sizeof cnt) == -1) {
+      return WSAGetLastError();
+    }
+
+    return 0;
   }
 
+  /* For those versions prior to Windows 10 version 1709,
+   * we fall back to SIO_KEEPALIVE_VALS that expects millisecond units.
+   * The SIO_KEEPALIVE_VALS IOCTL is supported on Windows 2000
```

Comment:
Perhaps we can drop this part since libuv supports much newer versions already.

Reply:
In that case, I assume function `minimal_windows10_version1709()` won't be needed anymore? I guess we can also drop the same runtime detection in `uv__tcp_try_connect`?

Reply:
Not really, we state we support "Windows 10", which means we cannot assume any minimum build.

Reply:
`TCP_KEEPIDLE` and `TCP_KEEPINTVL` are available starting with Windows 10, version 1709. `TCP_KEEPCNT` first appeared on Windows 10, version 1703.

I think we should use `SIO_KEEPALIVE_VALS` for Windows 10 versions before 1709?

Reply:
Yeah that sound about right.